Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -0,0 +1,638 @@
+
+  _simParticlesToKeep.clear();
+  _simPtrRemap.clear();
+
+  // Compress detector steps and record which SimParticles we want to keep
+  compressStrawGasSteps(event);
+  compressCaloShowerSteps(event);
+  compressCrvSteps(event);
+  if (_stepPointMCCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
+    // if we are not compressing StepPointMCs, then
+    // we want to make sure we record all their SimParticles
+    compressStepPointMCs(event);
+  }
+  if (_mcTrajectoryCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
+    // if we are not compressing MCTrajectories, then
+    // we want to make sure we record all their SimParticles
+    compressMCTrajectories(event);
+  }
+
+  // Compress the SimParticles and record their new keys
+  compressSimParticles(event);
+
+  // Now that we know which SimParticles we are keeping,
+  // we will keep the data products that are associated with these SimParticles
+  compressGenParticles();
+  if (_stepPointMCCompressionLevel == mu2e::CompressionLevel::kSimParticleCompression) {
+    // if we are compressing StepPointMCs based on the SimParticles we are keeping,
+    // then compressStepPointMCs now
+    compressStepPointMCs(event);
+  }
+  if (_mcTrajectoryCompressionLevel == mu2e::CompressionLevel::kSimParticleCompression) {
+    // if we are compressing MCTrajectories based on the SimParticles we are keeping,
+    // then compress MCTrajectories now
+    compressMCTrajectories(event);
+  }
+
+  // Update all the data products so that their SimParticlePtrs point to the new collection
+  updateStrawGasSteps();
+  updateCaloShowerSteps();
+  updateCrvSteps();
+  updateStepPointMCs();
+  updateMCTrajectories();
+
+  // Now add everything to the event
+  event.put(std::move(_newStrawGasSteps));
+  event.put(std::move(_newCaloShowerSteps));
+  event.put(std::move(_newCrvSteps));
+  for (const auto& i_tag : _conf.stepPointMCTags()) {
+    event.put(std::move(_newStepPointMCs.at(i_tag.instance())), i_tag.instance());
+  }
+  event.put(std::move(_newSimParticles));
+  event.put(std::move(_newGenParticles));
+  event.put(std::move(_newMCTrajectories));
+}
+
+void mu2e::CompressDetStepMCs::compressStrawGasSteps(const art::Event& event) {
+  art::Handle<mu2e::StrawGasStepCollection> strawGasStepsHandle;
+  event.getByLabel(_conf.strawGasStepTag(), strawGasStepsHandle);
+  const auto& strawGasSteps = *strawGasStepsHandle;
+  if(_conf.debugLevel()>0 && strawGasSteps.size()>0) {
+    std::cout << "Compressing StrawGasSteps from " << _conf.strawGasStepTag() << std::endl;
+  }
+  for (const auto& i_strawGasStep : strawGasSteps) {
+    if (_simParticleCompressionLevel == mu2e::CompressionLevel::kFullCompression) {
+      recordSimParticle(i_strawGasStep.simParticle());
+    }
+    StrawGasStep newStrawGasStep(i_strawGasStep);
+    _newStrawGasSteps->push_back(newStrawGasStep);
+  }
+  if(_strawGasStepCompressionLevel == mu2e::CompressionLevel::kNoCompression) {
```

Comment:
From looking at the code above,  newStrawGasSteps must always have the same length as strawGasSteps, so why is this test necessary?  And what is the use case for skipping the test (setting a higher compression level)?  Same comment for the other steps.

Reply:
This test is just a sanity check in case I add code that inadvertantly remove steps. I don't see a use case for compressing steps in this module. I'm thinking in case we want to refactor CompressDigiMCs to have a similar structure, then there would be some consistency between different compression modules. I also like the idea of being explicit in the configuration about which data products we are touching.